### ISSUE TITLE
MOE training Loss inconsistent after resume from old checkpoint

### ISSUE THREAD

**Issue by guozhen1997** [NVIDIA/Megatron-LM#761]
**MOE training Loss inconsistent after resume from old checkpoint**
Experimental conditions:

- the latest main branch
- use mcore
- expert-model-parallel-size > 1

<img width="928" alt="training loss" src="https://github.com/NVIDIA/Megatron-LM/assets/37097045/a605988d-9e5b-44ec-acda-60211e6bde9e">

The black line runs continuously and saves a ckpt every 100 steps.
The blue line is at the 100th step. ckpt is loaded and continues running.
What might cause this and how to fix it？

**Comment by fanshiqing** [comment]
Hi @guozhen1997 , we are also debugging on this issue. I will ping you when we find the root cause ASAP.

**Comment by fanshiqing** [comment]
Hi @guozhen1997 , this issue is caused by an incorrect implementation of the dual-optimizer state loading function, the fix MR is under review and will be published soon.

**Comment by binxuan** [comment]
Hi @fanshiqing , if we use the lagecy checkpointing method instead of the distributed checkpointing will we encounter this issue?

**Comment by fanshiqing** [comment]
Hi @guozhen1997 and @binxuan  , this issue has already been fixed by this [commit](https://github.com/NVIDIA/Megatron-LM/commit/1505db4cc4e9e94ee22583c76f7e425ea34f5aea).
